# ANN vs CNN on Cifar10

In [7]:
import tensorflow as tf
from tensorflow import keras
import numpy as np
from sklearn.model_selection import train_test_split

In [8]:
tf.config.experimental.list_physical_devices()

[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU')]

In [2]:
(x_train, y_train), (x_test, y_test) = tf.keras.datasets.cifar10.load_data()

In [3]:
x_train.shape

(50000, 32, 32, 3)

In [4]:

x_train = x_train.astype('float32') / 255
x_test = x_test.astype('float32') / 255

In [5]:
x_train[0].shape

(32, 32, 3)

In [6]:
x_train, x_val, y_train, y_val = train_test_split(x_train, y_train, test_size=0.2, random_state=42)

# $ 1. Artificial Neural Network $

In [32]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(32, 32, 3)),
    keras.layers.Dense(128, activation='relu'),
    keras.layers.Dense(64, activation='relu'),
    keras.layers.Dense(10, activation='sigmoid')
])

model.compile(optimizer='adam',
                loss='sparse_categorical_crossentropy',
                metrics=['accuracy'])

model.fit(x_train, y_train, validation_data=(x_val, y_val), epochs=10)

Epoch 1/10
1250/1250 [==============================] - 3s 2ms/step - loss: 1.9140 - accuracy: 0.3083 - val_loss: 1.8072 - val_accuracy: 0.3407
Epoch 2/10
1250/1250 [==============================] - 4s 3ms/step - loss: 1.7429 - accuracy: 0.3740 - val_loss: 1.7071 - val_accuracy: 0.3908
Epoch 3/10
1250/1250 [==============================] - 4s 3ms/step - loss: 1.6705 - accuracy: 0.4000 - val_loss: 1.6401 - val_accuracy: 0.4114
Epoch 4/10
1250/1250 [==============================] - 4s 3ms/step - loss: 1.6216 - accuracy: 0.4206 - val_loss: 1.6305 - val_accuracy: 0.4105
Epoch 5/10
1250/1250 [==============================] - 4s 3ms/step - loss: 1.5772 - accuracy: 0.4341 - val_loss: 1.6081 - val_accuracy: 0.4231
Epoch 6/10
1250/1250 [==============================] - 4s 3ms/step - loss: 1.5547 - accuracy: 0.4420 - val_loss: 1.5977 - val_accuracy: 0.4301
Epoch 7/10
1250/1250 [==============================] - 4s 3ms/step - loss: 1.5282 - accuracy: 0.4545 - val_loss: 1.5489 - val_accuracy:

In [34]:
test_loss, test_acc = model.evaluate(x_test, y_test, verbose=2)
print('Test accuracy:', test_acc)

313/313 - 0s - loss: 1.5078 - accuracy: 0.4665 - 339ms/epoch - 1ms/step
Test accuracy: 0.46650001406669617


# $ 2. Convolutional Neural Network $

In [12]:

model = keras.models.Sequential([
    keras.layers.Conv2D(32, (3, 3), activation='relu', input_shape=(32, 32, 3)),
    keras.layers.MaxPooling2D((2, 2)),
    keras.layers.Conv2D(64, (3, 3), activation='relu'),
    keras.layers.MaxPooling2D((2, 2)),
    keras.layers.Flatten(),
    keras.layers.Dense(64, activation='relu'),
    keras.layers.Dense(10, activation='softmax')
])


model.compile(optimizer='adam',
                loss='sparse_categorical_crossentropy',
                metrics=['accuracy'])

model.fit(x_train, y_train, validation_data=(x_val, y_val), epochs=10)


Epoch 1/10
1250/1250 [==============================] - 26s 20ms/step - loss: 1.5433 - accuracy: 0.4399 - val_loss: 1.3982 - val_accuracy: 0.4921
Epoch 2/10
1250/1250 [==============================] - 26s 21ms/step - loss: 1.1965 - accuracy: 0.5790 - val_loss: 1.1718 - val_accuracy: 0.5803
Epoch 3/10
1250/1250 [==============================] - 31s 25ms/step - loss: 1.0530 - accuracy: 0.6331 - val_loss: 1.0127 - val_accuracy: 0.6440
Epoch 4/10
1250/1250 [==============================] - 32s 26ms/step - loss: 0.9576 - accuracy: 0.6683 - val_loss: 1.0091 - val_accuracy: 0.6467
Epoch 5/10
1250/1250 [==============================] - 27s 22ms/step - loss: 0.8925 - accuracy: 0.6898 - val_loss: 0.9985 - val_accuracy: 0.6509
Epoch 6/10
1250/1250 [==============================] - 25s 20ms/step - loss: 0.8371 - accuracy: 0.7096 - val_loss: 0.9439 - val_accuracy: 0.6737
Epoch 7/10
1250/1250 [==============================] - 28s 22ms/step - loss: 0.7840 - accuracy: 0.7281 - val_loss: 0.9671 -

In [13]:
test_loss, test_acc = model.evaluate(x_test, y_test, verbose=2)
print('Test accuracy:', test_acc)

313/313 - 1s - loss: 0.9358 - accuracy: 0.6881 - 1s/epoch - 4ms/step
Test accuracy: 0.6880999803543091


# $ Analysis $

 . | Artificial Neural Network | Convolutional Neural Network
---|---|----
Accuracy | 0.466 | 0.688


# $ Efficiency $

    CNN is more efficient than ANN in terms of accuracy and time taken to train the model.

# $ Simplicity $

    Although ANN is Simple to implement, An ANN treats the whole image as a single entity and CNN treats the image as a collection of small entities. 
    